Ensure all ServiceWorkerRegistration files and directories are removed when registrations are cleared

szewai opened the pull request:
#### 05f2eb13cce60033cab4180a158f77677ff79cba
<pre>
Ensure all ServiceWorkerRegistration files and directories are removed when registrations are cleared
<a href="https://bugs.webkit.org/show_bug.cgi?id=257857">https://bugs.webkit.org/show_bug.cgi?id=257857</a>
rdar://110475749

Reviewed by Youenn Fablet.

This patch ensures -shm file, -wal file and empty Script folder are deleted when clearing ServiceWorker registrations,
so that ServiceWorker directory can become empty and also get deleted.

Test: WKWebsiteDataStore.RemoveServiceWorkerData

* Source/WebCore/workers/service/server/SWRegistrationDatabase.cpp:
(WebCore::scriptDirectoryPath):
(WebCore::scriptVersionDirectoryPath):
(WebCore::SWRegistrationDatabase::scriptStorage):
(WebCore::SWRegistrationDatabase::importRegistrations):
(WebCore::SWRegistrationDatabase::clearAllRegistrations):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebsiteDataStoreCustomPaths.mm:

Canonical link: <a href="https://commits.webkit.org/264998@main">https://commits.webkit.org/264998@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1bd7fa0c9fecc59f9063cfb8f80d6ea8bc28caf0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/9400 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/9677 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/9908 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/11062 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/9272 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/9409 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/11664 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/9639 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/12153 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/9551 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/10464 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/8076 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/11219 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/7761 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/8578 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/16008 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/8858 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/8727 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/12056 "12 api tests failed or timed out") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/9223 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/7503 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/8435 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2277 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/12660 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/8985 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/workers/service/server/SWRegistrationDatabase.cpp:
```diff
@@ -79,7 +79,16 @@ static String scriptDirectoryPath(const String& directory)
+    if (scriptDirectory.isEmpty())
+        return emptyString();
+
+    return FileSystem::pathByAppendingComponent(scriptDirectory, scriptVersion);
```
It is doing a bit more than just removing she and wal files.
We will now delete ./Scripts folder instead of ./Scripts/V1 folder.
This is probably fine but it is worth mentioning this in the commit message.

szewai replied:
`We will now delete ./Scripts folder instead of ./Scripts/V1 folder.`
This is expected, I should say "delete shm file, wal file, and empty Scripts folder"; will add

youennf commented on Source/WebCore/workers/service/server/SWRegistrationDatabase.cpp:
```diff
@@ -296,8 +305,10 @@ std::optional<Vector<ServiceWorkerContextData>> SWRegistrationDatabase::importRe
 
-    if (!m_database)
+    if (!m_database) {
+        clearAllRegistrations();
```
I am not clear why we are adding this change.
This happens in the case where prepareDatabase returned true and m_database is nullptr.
This means that the database file does not exist. In that case, clearAllRegistrations is almost a no-op, no?

szewai replied:
This patch ensures clearing registrations will delete all related files/folders. But for clients use build with this issue, they might be left in a state where they have a ServiceWorker directory (and some files inside) but no valid ServiceWorker registrations (since they are cleared), and there is no way for client to delete them.

The `clearAllRegistrations()` here basically performs a cleanup for this issue.

youennf replied:
Clients can call [websiteDataStore removeDataOfTypes] if they want to remove all the files in that folder without this change, right?
If so, this change does not seem necessary for fixing the bug.
Also, it does not seem covered by the test.

In general, I am not sure calling importRegistrations() should trigger clearing the folder.

szewai replied:
> Clients can call [websiteDataStore removeDataOfTypes] if they want to remove all the files in that folder without this change, right?

hmm not really. We check if registration exists before we remove the data; if there is no registration, then we don't remove it. In this case, if there is only shm file left in the folder, it will be kept after `removeDataOfTypes`, but user will not see the site when `fetchDataRecordsOfTypes`.
We do the "fetch-then-delete" to avoid deleting files that are not belonged to WebKit.

> In general, I am not sure calling importRegistrations() should trigger clearing the folder.

I choose `importRegistrations` because it is currently called only once at session creation time. We can add the cleanup function elsewhere.

szewai replied:
I will land this patch as it is now, so we could fix the slowness issue on EWS bots: too many origin directories are not deleted as expected, and iterating all of them (during data fetch/removal) takes so long that API tests are timed out

I could make followup if you have other suggestion for this cleanup @youennf

youennf commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WebsiteDataStoreCustomPaths.mm:
```diff
@@ -1714,3 +1714,58 @@ function install()
+    done = false;
+    [websiteDataStore removeDataOfTypes:[NSSet setWithObjects:WKWebsiteDataTypeServiceWorkerRegistrations, nil] modifiedSince:[NSDate distantPast] completionHandler:^() {
+        done = true;
+        EXPECT_FALSE([[NSFileManager defaultManager] fileExistsAtPath:serviceWorkerDirectoryString]);
```
I would have moved it outside the block to align with the previous EXPEC_TRUE(file is there).

szewai committed:
Ensure all ServiceWorkerRegistration files and directories are removed when registrations are cleared
https://bugs.webkit.org/show_bug.cgi?id=257857
rdar://110475749

Reviewed by Youenn Fablet.

This patch ensures -shm file, -wal file and empty Script folder are deleted when clearing ServiceWorker registrations,
so that ServiceWorker directory can become empty and also get deleted.

Test: WKWebsiteDataStore.RemoveServiceWorkerData

* Source/WebCore/workers/service/server/SWRegistrationDatabase.cpp:
(WebCore::scriptDirectoryPath):
(WebCore::scriptVersionDirectoryPath):
(WebCore::SWRegistrationDatabase::scriptStorage):
(WebCore::SWRegistrationDatabase::importRegistrations):
(WebCore::SWRegistrationDatabase::clearAllRegistrations):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebsiteDataStoreCustomPaths.mm:

Canonical link: https://commits.webkit.org/264998@main
